### ISSUE TITLE
Unify reshape routing to make extension easier

### ISSUE THREAD

**Issue by darsnack** [JuliaLang/julia#39123]
We ran into an issue trying to extend `Base.reshape` in https://github.com/FluxML/Flux.jl/pull/1448. Ideally, someone should only have to overload
```julia
Base.reshape(x::MySpecialArray, dims::Dims)
```
to get insert their custom behavior. Currently, calls involving a single `:` [bypass this method signature](https://github.com/JuliaLang/julia/blob/788b2c77c10c2160f4794a4d4b6b81a95a90940c/base/reshapedarray.jl#L118) directly to the internal `_reshape`. I'd like to redefine this line to be:
```julia
reshape(parent::AbstractArray, dims::Tuple{Vararg{Union{Int,Colon}}}) = reshape(parent, _reshape_uncolon(parent, dims))
```
so that it routes through the signature above after the uncolon logic is applied. I see no real problems with this, but I wanted to create an issue before a PR to get told why I'm wrong.

**Comment by mbauman** [comment]
This is actually something I've wanted to do for a long time.  It's a bit of a rats nest right now, and IIRC there are two levels of redirection — one for colons and one for axes/offset arrays.  It's the latter that poses the bigger problem.

**Comment by darsnack** [comment]
I assume the second level is [this](https://github.com/JuliaLang/julia/blob/788b2c77c10c2160f4794a4d4b6b81a95a90940c/base/reshapedarray.jl#L177)? So would you want to lift that logic out of an internal function?

**Comment by mbauman** [comment]
Yes, I'd think so.  But it's a general problem with Offset arrays — general array types want the "extension" method to be the simple `::Dims`, whereas offset arrays require `::Axes`.  It makes things go in circles.

**Comment by darsnack** [comment]
Okay I see why there was this extra path now. There's a dispatch ambiguity if you redefine it like I want mentioned above. Maybe we can just tweak [`_reshape_uncolon`](https://github.com/JuliaLang/julia/blob/788b2c77c10c2160f4794a4d4b6b81a95a90940c/base/reshapedarray.jl#L119) to no-op on the no colon case? Then we just remove `reshape(parent, ::Dims)`.

**Comment by darsnack** [comment]
It looks like [there was already a fix](https://github.com/JuliaLang/julia/commit/0f0bc294f7b18d685c4998ada9c2c33841a74f2d) for the initial complaint. I can still keep this issue open since we still want to clean up the interface.
